mypy/types.py

```diff
@@ -3381,12 +3381,20 @@ def _flattened(types: Iterable[Type]) -> Iterable[Type]:
 
 
 def flatten_nested_unions(
-    types: Iterable[Type], handle_type_alias_type: bool = True
+    types: Sequence[Type], handle_type_alias_type: bool = True
 ) -> list[Type]:
     """Flatten nested unions in a type list."""
+    if not isinstance(types, list):
+        typelist = list(types)
+    else:
+        typelist = cast("list[Type]", types)
+
+    # Fast path: most of the time there is nothing to flatten
+    if not any(isinstance(t, (TypeAliasType, UnionType)) for t in typelist):  # type: ignore[misc]
+        return typelist
```

Comment:
I think we only need to convert it to a `list` if we're taking the fast path, correct?

```suggestion
    # Fast path: most of the time there is nothing to flatten
    if not any(isinstance(t, (TypeAliasType, UnionType)) for t in types):  # type: ignore[misc]
        if not isinstance(types, list):
            return list(types)
        return cast("list[Type]", types)
```

Reply:
I'm first converting it to a list since iterating over a list if faster than iterating over a sequence, and since we need to return a list anyway, so it would only delay the construction of the list. Since we take the fast path the vast majority of the time, the redundant construction of a list in the slow path is not a big deal, and we save some of the cost through faster iteration. If mypyc was smarter about iterating over sequences, your suggested approach would make sense.

Reply:
Thanks for the explanation!